base/abstractarray.jl

```diff
@@ -126,6 +126,34 @@ _length(A) = length(A)
 endof(a::AbstractArray) = length(a)
 first(a::AbstractArray) = a[first(eachindex(a))]
 
+function fixate_eltype{T<:AbstractArray}(::Type{T}, default::Type)
+    @_pure_meta
+    aa_type = T
+    while aa_type.name.primary != AbstractArray
+        aa_type = supertype(aa_type)
+    end
+    if isa(aa_type.parameters[1], TypeVar)
+        varname = aa_type.parameters[1].name
+        new_params = collect(T.parameters)
+        for i in eachindex(new_params)
+            p = new_params[i]
+            if isa(p, TypeVar) && p.name == varname
+                new_params[i] = default
+            end
+        end
+        return T.name.primary{new_params...}
+    else
+        return T
+    end
+end
```

Comment:
Love the idea. Is this intended to give `fixate_eltype(Diagonal{Int}, Float64) === Float64`? If that is the case, I think there is a problem with the implementation, an alternative would be

```julia
function promote_containereltype{A<:AbstractArray}(::Type{A}, T::Type)
    Base.@_pure_meta
    AA = A
    AT = AA.name.primary
    while AT.name.primary != AbstractArray
        AA = supertype(AA)
        AT = supertype(AT)
    end
    if AA.parameters[1] === T
        return A
    else
        tvarname = AT.parameters[1].name
        params = A.parameters
        new_params = collect(A.name.primary.parameters)
        for i in eachindex(new_params)
            p = new_params[i]
            if isa(p, TypeVar) && p.name == tvarname
                new_params[i] = T
            else
                new_params[i] = params[i]
            end
        end
        return A.name.primary{new_params...}
    end
end
```